File: android/app/src/main/java/com/viant/agently/android/ReportRuntimeExportHandler.kt
package com.viant.agently.android

import com.viant.agentlysdk.AgentlyClient
import com.viant.agentlysdk.DownloadFileOutput
import com.viant.agentlysdk.GeneratedFileEntry
import com.viant.forgeandroid.runtime.ForgeRuntime
import com.viant.forgeandroid.runtime.JsonUtil
import com.viant.forgeandroid.runtime.snapshotFeedDataSources
import kotlinx.coroutines.delay
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonPrimitive
import java.util.Base64

private const val REPORT_EXPORT_POLL_INTERVAL_MS = 1_500L
private const val REPORT_EXPORT_POLL_ATTEMPTS = 20

internal fun registerReportRuntimeExportHandler(
    forgeRuntime: ForgeRuntime,
    client: AgentlyClient,
    conversationIdProvider: () -> String?,
    onError: (String?) -> Unit,
    openPdf: (GeneratedFileEntry, DownloadFileOutput) -> Boolean
) {
    forgeRuntime.registerHandler("reportRuntime.exportPdf") { args ->
        val exportRequest = JsonUtil.asStringMap(args.args["exportRequest"]).takeIf { it.isNotEmpty() }
        if (exportRequest == null) {
            onError("No report export request is available.")
            return@registerHandler false
        }
        val conversationId = conversationIdProvider()?.trim().orEmpty()
        val artifact = try {
            exportReportRuntimePdf(
                client = client,
                exportRequest = exportRequest,
                conversationId = conversationId
            )
        } catch (err: Throwable) {
            onError(reportRuntimeExportErrorMessage(err))
            return@registerHandler false
        }
        val opened = openPdf(artifact.file, artifact.downloaded)
        if (!opened) {
            onError("PDF export completed, but no PDF viewer was available.")
            return@registerHandler false
        }
        onError(null)
        true
    }
    forgeRuntime.registerHandler("feed.print") { args ->
        val context = args.context
        if (context == null) {
            onError("No active feed is available for PDF export.")
            return@registerHandler false
        }
        val state = forgeRuntime.windowState(context.window.windowId)
        val feedId = state?.windowKey?.removePrefix("feed-")
            ?.removeSuffix("-${state.conversationId.orEmpty()}")
            ?.takeIf(String::isNotBlank)
            ?: "feed"
        val exportRequest = buildFeedReportExportRequest(
            forgeRuntime = forgeRuntime,
            windowId = context.window.windowId,
            feedId = feedId,
            title = state?.windowTitle ?: "Tool feed"
        )
        val conversationId = conversationIdProvider()?.trim().orEmpty()
        val artifact = try {
            exportReportRuntimePdf(client, exportRequest, conversationId)
        } catch (err: Throwable) {
            onError(reportRuntimeExportErrorMessage(err))
            return@registerHandler false
        }
        val opened = openPdf(artifact.file, artifact.downloaded)
        if (!opened) {
            onError("PDF export completed, but no PDF viewer was available.")
            return@registerHandler false
        }
        onError(null)
        true
    }
}

internal fun buildFeedReportExportRequest(
    forgeRuntime: ForgeRuntime,
    windowId: String,
    feedId: String,
    title: String
): Map<String, Any?> {
    val metadata = forgeRuntime.metadataSignal(windowId).peek()
        ?: error("Feed PDF export requires window metadata.")
    val safeFeedId = safeFeedReportId(feedId)
    val snapshots = snapshotFeedDataSources(
        forgeRuntime.windowContext(windowId),
        metadata.dataSources.keys
    )
    val overrides = snapshots.mapValues { (_, snapshot) ->
        mapOf(
            "form" to snapshot.form,
            "collection" to snapshot.collection,
            "selection" to snapshot.selection
        )
    }
    return mapOf(
        "viewRef" to "feed://$feedId",
        "format" to "pdf",
        "reportId" to "feed_$safeFeedId",
        "title" to title,
        "dataSourceRefs" to metadata.dataSources.keys.sorted(),
        "dataSourceOverrides" to overrides,
        "target" to mapOf(
            "platform" to forgeRuntime.targetContext.platform,
            "formFactor" to forgeRuntime.targetContext.formFactor,
            "surface" to forgeRuntime.targetContext.surface,
            "capabilities" to forgeRuntime.targetContext.capabilities.sorted()
        )
    )
}

private fun safeFeedReportId(value: String): String = value.trim().lowercase()
    .replace(Regex("[^a-z0-9]+"), "_")
    .trim('_')
    .ifBlank { "feed" }

internal data class ReportRuntimeExportArtifact(
    val file: GeneratedFileEntry,
    val downloaded: DownloadFileOutput
)

internal suspend fun exportReportRuntimePdf(
    client: AgentlyClient,
    exportRequest: Map<String, Any?>,
    conversationId: String = ""
): ReportRuntimeExportArtifact {
	val viewRef = stringValue(exportRequest["viewRef"])
	val fences = JsonUtil.anyToElement(exportRequest["fences"]) as? JsonArray
	val completedJob = if (viewRef.isNotBlank()) {
		val args = exportRequest.mapValues { JsonUtil.anyToElement(it.value) }.toMutableMap().apply {
			put("conversationId", JsonPrimitive(conversationId))
		}
		val result = executeReportingToolObject(
			client = client,
			toolName = "reporting:compile_and_export_forge_ui",
			args = args,
			conversationId = conversationId
		)
		val job = normalizeReportExportJob(result["job"] as? JsonObject ?: JsonObject(emptyMap()))
		val artifactId = stringValue((result["artifact"] as? JsonObject)?.get("artifactId"))
		job.copy(artifactId = job.artifactId.ifBlank { artifactId })
	} else if (fences != null && fences.isNotEmpty()) {
		val result = executeReportingToolObject(
			client = client,
			toolName = "reporting:compile_and_export_fenced_report",
			args = mapOf(
				"reportId" to JsonPrimitive(stringValue(exportRequest["reportId"])),
				"format" to JsonPrimitive("pdf"),
				"fences" to fences,
				"conversationId" to JsonPrimitive(conversationId)
			),
			conversationId = conversationId
		)
		val job = normalizeReportExportJob(result["job"] as? JsonObject ?: JsonObject(emptyMap()))
		val artifactId = stringValue((result["artifact"] as? JsonObject)?.get("artifactId"))
		job.copy(artifactId = job.artifactId.ifBlank { artifactId })
	} else {
		val submitResult = executeReportingToolObject(
			client = client,
			toolName = "reporting:submit_export",
			args = normalizeReportRuntimeExportRequest(exportRequest),
			conversationId = conversationId
		)
		waitForReportExportArtifact(
			client = client,
			initialJob = normalizeReportExportJob(submitResult),
			conversationId = conversationId
		)
	}
    val artifactId = completedJob.artifactId.ifBlank {
        error("Report PDF export completed without an artifact id.")
    }
    val artifact = executeReportingToolObject(
        client = client,
        toolName = "reporting:get_artifact",
        args = mapOf(
            "artifactId" to JsonPrimitive(artifactId),
            "includeData" to JsonPrimitive(true)
        ),
        conversationId = conversationId
    )
    val bytes = decodeReportArtifactBytes(artifact)
    if (bytes.isEmpty()) {
        error("Report PDF artifact was empty.")
    }
    val title = stringValue(exportRequest["title"]).ifBlank { completedJob.artifactRef.ifBlank { artifactId } }
    val name = sanitizeDownloadFileName(
        stringValue(artifact["filename"])
            .ifBlank { stringValue(artifact["name"]) }
            .ifBlank { "$title.pdf" }
    ).ifBlank { "$artifactId.pdf" }
    val contentType = stringValue(artifact["contentType"]).ifBlank { "application/pdf" }
    return ReportRuntimeExportArtifact(
        file = GeneratedFileEntry(
            id = artifactId,
            conversationId = conversationId.ifBlank { null },
            filename = name,
            mimeType = contentType,
            sizeBytes = bytes.size
        ),
        downloaded = DownloadFileOutput(
            name = name,
            contentType = contentType,
            data = bytes
        )
    )
}

private fun normalizeReportRuntimeExportRequest(exportRequest: Map<String, Any?>): Map<String, JsonElement> {
    val artifactRef = stringValue(exportRequest["artifactRef"])
        .takeIf { it.isNotBlank() }
        ?: error("Report PDF export requires artifactRef.")
    val reportSpec = JsonUtil.anyToElement(exportRequest["reportSpec"]) as? JsonObject
        ?: error("Report PDF export requires reportSpec.")
    val reportFill = JsonUtil.anyToElement(exportRequest["reportFill"]) as? JsonObject
        ?: error("Report PDF export requires reportFill.")
    val reportPrint = JsonUtil.anyToElement(exportRequest["reportPrint"]) as? JsonObject
        ?: error("Report PDF export requires reportPrint.")
    return linkedMapOf(
        "artifactRef" to JsonPrimitive(artifactRef),
        "format" to JsonPrimitive("pdf"),
        "scope" to JsonPrimitive("draft"),
        "reportSpec" to reportSpec,
        "reportFill" to reportFill,
        "reportPrint" to reportPrint
    )
}

private suspend fun waitForReportExportArtifact(
    client: AgentlyClient,
    initialJob: ReportExportJobState,
    conversationId: String
): ReportExportJobState {
    var job = initialJob
    repeat(REPORT_EXPORT_POLL_ATTEMPTS) { attempt ->
        if (job.artifactId.isNotBlank()) {
            return job
        }
        val status = job.status.lowercase()
        if (status in setOf("failed", "canceled", "cancelled")) {
            error(job.error.ifBlank { "Report PDF export failed." })
        }
        if (job.jobId.isBlank()) {
            error("Report PDF export did not return a job id.")
        }
        if (attempt > 0) {
            delay(REPORT_EXPORT_POLL_INTERVAL_MS)
        }
        val statusResult = executeReportingToolObject(
            client = client,
            toolName = "reporting:get_export_status",
            args = mapOf("jobId" to JsonPrimitive(job.jobId)),
            conversationId = conversationId
        )
        job = normalizeReportExportJob(statusResult)
    }
    error("Report PDF export did not finish in time.")
}

private data class ReportExportJobState(
    val jobId: String,
    val artifactId: String,
    val artifactRef: String,
    val status: String,
    val error: String
)

private fun normalizeReportExportJob(value: JsonObject): ReportExportJobState {
    return ReportExportJobState(
        jobId = stringValue(value["jobId"]),
        artifactId = stringValue(value["artifactId"]),
        artifactRef = stringValue(value["artifactRef"]),
        status = stringValue(value["status"]),
        error = stringValue(value["error"])
    )
}

private suspend fun executeReportingToolObject(
    client: AgentlyClient,
    toolName: String,
    args: Map<String, JsonElement>,
    conversationId: String
): JsonObject {
    val raw = client.executeTool(
        name = toolName,
        args = args,
        conversationId = conversationId.ifBlank { null }
    )
    if (raw.isBlank()) {
        error("Reporting tool $toolName returned an empty response.")
    }
    val element = Json.parseToJsonElement(raw)
    return element as? JsonObject
        ?: error("Reporting tool $toolName returned an unexpected response.")
}

private fun decodeReportArtifactBytes(artifact: JsonObject): ByteArray {
    val data = stringValue(artifact["data"])
    if (data.isNotBlank()) {
        return Base64.getDecoder().decode(data)
    }
    val bytes = artifact["bytes"] as? JsonArray ?: return ByteArray(0)
    return bytes.mapNotNull { element ->
        element.jsonPrimitive.contentOrNull?.toIntOrNull()?.toByte()
    }.toByteArray()
}

private fun stringValue(value: Any?): String {
    return when (value) {
        null -> ""
        is String -> value.trim()
        is JsonPrimitive -> value.contentOrNull?.trim().orEmpty()
        is JsonElement -> (value as? JsonPrimitive)?.contentOrNull?.trim().orEmpty()
        else -> value.toString().trim()
    }
}

internal fun reportRuntimeExportErrorMessage(error: Throwable): String {
    val raw = error.message.orEmpty()
    val serverDetail = raw.indexOf('{')
        .takeIf { it >= 0 }
        ?.let { start ->
            runCatching {
                val response = Json.parseToJsonElement(raw.substring(start)) as? JsonObject
                (response?.get("error") as? JsonPrimitive)?.contentOrNull
            }.getOrNull()
        }
        ?.replace("\n", " ")
        ?.trim()
	val detail = serverDetail
		?.trim()
		?.takeIf { it.isNotBlank() }
    return if (detail == null) {
        "Unable to create the report PDF. Please try again."
    } else {
        "Unable to create the report PDF: $detail"
    }
}

File: ios/Sources/AgentlyAppFoundation/App/ReportRuntimeExportHandler.swift
import Foundation
import AgentlySDK
import ForgeIOSRuntime

private let reportExportPollIntervalNanoseconds: UInt64 = 1_500_000_000
private let reportExportPollAttempts = 20

typealias SDKJSONValue = AgentlySDK.JSONValue

struct ReportRuntimeExportArtifact: Sendable, Equatable {
    let id: String
    let name: String
    let contentType: String
    let data: Data
}

enum ReportRuntimeExportError: Error, LocalizedError {
    case missingExportRequest
    case emptyToolResult(String)
    case unexpectedToolResult(String)
    case missingJobID
    case failed(String)
    case timedOut
    case missingArtifactID
    case emptyArtifact

    var errorDescription: String? {
        switch self {
        case .missingExportRequest:
            return "No report export request is available."
        case .emptyToolResult(let tool):
            return "Reporting tool \(tool) returned an empty response."
        case .unexpectedToolResult(let tool):
            return "Reporting tool \(tool) returned an unexpected response."
        case .missingJobID:
            return "Report PDF export did not return a job id."
        case .failed(let message):
            return message.isEmpty ? "Report PDF export failed." : message
        case .timedOut:
            return "Report PDF export did not finish in time."
        case .missingArtifactID:
            return "Report PDF export completed without an artifact id."
        case .emptyArtifact:
            return "Report PDF artifact was empty."
        }
    }
}

func registerReportRuntimeExportHandler(
    forgeRuntime: ForgeRuntime,
    client: AgentlyClient,
    conversationIDProvider: @escaping @Sendable () async -> String?,
    onComplete: @escaping @Sendable (ReportRuntimeExportArtifact) async -> Void,
    onError: @escaping @Sendable (String?) async -> Void
) async {
    await forgeRuntime.registerHandler("reportRuntime.exportPdf") { args in
        guard let exportRequest = args.args["exportRequest"]?.objectValue, !exportRequest.isEmpty else {
            await onError(ReportRuntimeExportError.missingExportRequest.localizedDescription)
            return .bool(false)
        }
        do {
            let artifact = try await exportReportRuntimePDF(
                client: client,
                exportRequest: exportRequest,
                conversationID: await conversationIDProvider()?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            )
            await onComplete(artifact)
            await onError(nil)
            return .bool(true)
        } catch {
            await onError(reportRuntimeExportErrorMessage(error))
            return .bool(false)
        }
    }
}

func reportRuntimeExportErrorMessage(_ error: Error) -> String {
    let raw = error.localizedDescription
    let serverDetail: String? = {
        guard let start = raw.firstIndex(of: "{") else { return nil }
        let suffix = String(raw[start...])
        guard let data = suffix.data(using: .utf8),
              let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let message = object["error"] as? String else {
            return nil
        }
        let detail = message
            .replacingOccurrences(of: "\n", with: " ")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        return detail
            .replacingOccurrences(of: "^reporting export:\\s*", with: "", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .nonEmpty
    }()
    let diagnostic = (serverDetail ?? raw).lowercased()
    if diagnostic.contains("scratchpad") ||
        diagnostic.contains("storage.googleapis.com") ||
        diagnostic.contains("unable to generate access token") {
        return "The PDF was created, but report storage is temporarily unavailable. Please try again."
    }
    guard let serverDetail else {
        return "Unable to create the report PDF. Please try again."
    }
    return "Unable to create the report PDF: \(serverDetail)"
}

func exportReportRuntimePDF(
    client: AgentlyClient,
    exportRequest: [String: ForgeJSONValue],
    conversationID: String = ""
) async throws -> ReportRuntimeExportArtifact {
    let completedJob: ReportExportJobState
    if let fences = exportRequest["fences"]?.arrayValue, !fences.isEmpty {
        let compiled = try await executeReportingToolObject(
            client: client,
            toolName: "reporting:compile_and_export_fenced_report",
            args: [
                "reportId": .string(stringValue(exportRequest["reportId"])),
                "format": .string("pdf"),
                "fences": .array(fences.map(sdkJSONValue(from:))),
                "conversationId": .string(conversationID)
            ],
            conversationID: conversationID
        )
        let job = normalizeReportExportJob(compiled["job"]?.objectValue ?? [:])
        let artifactID = stringValue(compiled["artifact"]?.objectValue?["artifactId"])
        completedJob = ReportExportJobState(
            jobID: job.jobID,
            artifactID: job.artifactID.nonEmpty ?? artifactID,
            artifactRef: job.artifactRef,
            status: job.status,
            error: job.error
        )
    } else {
        let request = normalizeReportRuntimeExportRequest(exportRequest)
        let submitResult = try await executeReportingToolObject(
            client: client,
            toolName: "reporting:submit_export",
            args: request,
            conversationID: conversationID
        )
        completedJob = try await waitForReportExportArtifact(
            client: client,
            initialJob: normalizeReportExportJob(submitResult),
            conversationID: conversationID
        )
    }
    guard !completedJob.artifactID.isEmpty else {
        throw ReportRuntimeExportError.missingArtifactID
    }
    let artifact = try await executeReportingToolObject(
        client: client,
        toolName: "reporting:get_artifact",
        args: [
            "artifactId": .string(completedJob.artifactID),
            "includeData": .bool(true)
        ],
        conversationID: conversationID
    )
    let bytes = decodeReportArtifactBytes(artifact)
    guard !bytes.isEmpty else {
        throw ReportRuntimeExportError.emptyArtifact
    }
    let title = stringValue(exportRequest["title"]).nonEmpty ?? completedJob.artifactRef.nonEmpty ?? completedJob.artifactID
    let name = sanitizeReportRuntimeExportFilename(
        stringValue(artifact["filename"]).nonEmpty
            ?? stringValue(artifact["name"]).nonEmpty
            ?? "\(title).pdf"
    )
    let contentType = stringValue(artifact["contentType"]).nonEmpty ?? "application/pdf"
    return ReportRuntimeExportArtifact(
        id: completedJob.artifactID,
        name: name.isEmpty ? "\(completedJob.artifactID).pdf" : name,
        contentType: contentType,
        data: bytes
    )
}

func downloadReportRuntimeArtifact(
    client: AgentlyClient,
    artifactID: String,
    conversationID: String = ""
) async throws -> ReportRuntimeExportArtifact {
    let normalizedID = artifactID.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !normalizedID.isEmpty else {
        throw ReportRuntimeExportError.missingArtifactID
    }
    let artifact = try await executeReportingToolObject(
        client: client,
        toolName: "reporting:get_artifact",
        args: [
            "artifactId": .string(normalizedID),
            "includeData": .bool(true)
        ],
        conversationID: conversationID
    )
    let bytes = decodeReportArtifactBytes(artifact)
    guard !bytes.isEmpty else {
        throw ReportRuntimeExportError.emptyArtifact
    }
    let name = sanitizeReportRuntimeExportFilename(
        stringValue(artifact["filename"]).nonEmpty
            ?? stringValue(artifact["name"]).nonEmpty
            ?? "(normalizedID).pdf"
    )
    return ReportRuntimeExportArtifact(
        id: normalizedID,
        name: name,
        contentType: stringValue(artifact["contentType"]).nonEmpty ?? "application/pdf",
        data: bytes
    )
}

private func normalizeReportRuntimeExportRequest(_ exportRequest: [String: ForgeJSONValue]) -> [String: SDKJSONValue] {
    let artifactRef = stringValue(exportRequest["artifactRef"]).nonEmpty
        ?? "report://runtime/\(stringValue(exportRequest["title"]).nonEmpty ?? "report")"
    return [
        "artifactRef": .string(artifactRef),
        "format": .string("pdf"),
        "scope": .string("draft"),
        "reportSpec": sdkJSONValue(from: exportRequest["reportSpec"] ?? ForgeJSONValue.null),
        "reportFill": sdkJSONValue(from: exportRequest["reportFill"] ?? ForgeJSONValue.null),
        "reportPrint": sdkJSONValue(from: exportRequest["reportPrint"] ?? ForgeJSONValue.null)
    ]
}

private func waitForReportExportArtifact(
    client: AgentlyClient,
    initialJob: ReportExportJobState,
    conversationID: String
) async throws -> ReportExportJobState {
    var job = initialJob
    for attempt in 0..<reportExportPollAttempts {
        if !job.artifactID.isEmpty {
            return job
        }
        if ["failed", "canceled", "cancelled"].contains(job.status.lowercased()) {
            throw ReportRuntimeExportError.failed(job.error)
        }
        if job.jobID.isEmpty {
            throw ReportRuntimeExportError.missingJobID
        }
        if attempt > 0 {
            try await Task.sleep(nanoseconds: reportExportPollIntervalNanoseconds)
        }
        let statusResult = try await executeReportingToolObject(
            client: client,
            toolName: "reporting:get_export_status",
            args: ["jobId": .string(job.jobID)],
            conversationID: conversationID
        )
        job = normalizeReportExportJob(statusResult)
    }
    throw ReportRuntimeExportError.timedOut
}

private struct ReportExportJobState: Sendable, Equatable {
    let jobID: String
    let artifactID: String
    let artifactRef: String
    let status: String
    let error: String
}

private func normalizeReportExportJob(_ value: [String: SDKJSONValue]) -> ReportExportJobState {
    ReportExportJobState(
        jobID: stringValue(value["jobId"]),
        artifactID: stringValue(value["artifactId"]),
        artifactRef: stringValue(value["artifactRef"]),
        status: stringValue(value["status"]),
        error: stringValue(value["error"])
    )
}

private func executeReportingToolObject(
    client: AgentlyClient,
    toolName: String,
    args: [String: SDKJSONValue],
    conversationID: String
) async throws -> [String: SDKJSONValue] {
    let raw = try await client.executeTool(
        name: toolName,
        args: args,
        conversationID: conversationID.isEmpty ? nil : conversationID
    )
    guard !raw.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
        throw ReportRuntimeExportError.emptyToolResult(toolName)
    }
    let value = try JSONDecoder().decode(SDKJSONValue.self, from: Data(raw.utf8))
    guard case .object(let object) = value else {
        throw ReportRuntimeExportError.unexpectedToolResult(toolName)
    }
    return object
}

private func decodeReportArtifactBytes(_ artifact: [String: SDKJSONValue]) -> Data {
    let data = stringValue(artifact["data"])
    if !data.isEmpty, let decoded = Data(base64Encoded: data) {
        return decoded
    }
    guard case .array(let bytes)? = artifact["bytes"] else {
        return Data()
    }
    return Data(bytes.compactMap { value in
        guard case .number(let number) = value else { return nil }
        return UInt8(exactly: Int(number))
    })
}

private func sdkJSONValue(from value: ForgeJSONValue) -> SDKJSONValue {
    switch value {
    case .string(let string):
        return .string(string)
    case .number(let number):
        return .number(number)
    case .bool(let bool):
        return .bool(bool)
    case .array(let values):
        return .array(values.map(sdkJSONValue(from:)))
    case .object(let object):
        return .object(object.mapValues(sdkJSONValue(from:)))
    case .null:
        return .null
    }
}

private func stringValue(_ value: ForgeJSONValue?) -> String {
    guard let value else { return "" }
    switch value {
    case .string(let string):
        return string.trimmingCharacters(in: CharacterSet.whitespacesAndNewlines)
    case .number(let number):
        return String(number)
    case .bool(let bool):
        return bool ? "true" : "false"
    default:
        return ""
    }
}

private func stringValue(_ value: SDKJSONValue?) -> String {
    guard let value else { return "" }
    switch value {
    case .string(let string):
        return string.trimmingCharacters(in: .whitespacesAndNewlines)
    case .number(let number):
        return String(number)
    case .bool(let bool):
        return bool ? "true" : "false"
    default:
        return ""
    }
}

private func sanitizeReportRuntimeExportFilename(_ value: String) -> String {
    let base = value.trimmingCharacters(in: .whitespacesAndNewlines).nonEmpty ?? "report.pdf"
    let cleaned = base
        .replacingOccurrences(of: "/", with: "-")
        .replacingOccurrences(of: ":", with: "-")
    return cleaned.contains(".") ? cleaned : "\(cleaned).pdf"
}

private extension String {
    var nonEmpty: String? {
        let trimmed = trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }
}
